runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,726 @@
+---@return vim.snippet.Position
+local function cursor_position()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  return { cursor[1] - 1, cursor[2] }
+end
+
+---Return one indent string of current buffer.
+---@return string
+local function get_one_indent()
+  local expandtab = vim.api.nvim_get_option('expandtab')
+  if not expandtab then
+    return '\t'
+  end
+  local shiftwidth = vim.api.nvim_get_option('shiftwidth')
+  if shiftwidth > 0 then
+    return string.rep(' ', shiftwidth)
+  end
+  return string.rep(' ', vim.api.nvim_get_option('tabstop'))
+end
+
+---Return base indent of current line.
+---@return string
+local function get_base_indent()
+  return string.match(vim.api.nvim_get_current_line(), '^%s*') or ''
+end
+
+---Adjust snippet indent
+---@param snippet_text string
+---@return string
+local function adjust_indent(snippet_text)
+  local one_indent = get_one_indent()
+  local base_indent = get_base_indent()
+
+  local adjusted_snippet_text = {}
+  local snippet_lines = vim.split(snippet_text, '\n', { plain = true })
+  for i, line in ipairs(snippet_lines) do
+    if i == 1 then
+      -- Use first line as-is.
+      table.insert(adjusted_snippet_text, line)
+    elseif i ~= #snippet_lines and string.match(line, '^%s*$') then
+      -- Remove space only lines except the last line.
+      table.insert(adjusted_snippet_text, '')
+    else
+      -- 1. Change \t as one_indent.
+      -- 2. Add base_indent.
+      line = string.gsub(line, '^(\t+)', function(indent)
+        return string.gsub(indent, '\t', one_indent)
+      end)
+      table.insert(adjusted_snippet_text, base_indent .. line)
+    end
+  end
+  return table.concat(adjusted_snippet_text, '\n')
+end
+
+---Get range from extmark.
+---@param ns integer
+---@param mark_id vim.snippet.MarkId
+---@return vim.snippet.Range
+local function get_range_from_mark(ns, mark_id)
+  local mark = vim.api.nvim_buf_get_extmark_by_id(0, ns, mark_id, { details = true })
+  local s = { mark[1], mark[2] }
+  local e = { mark[3].end_row, mark[3].end_col }
+  if s[1] > e[1] or (s[1] == e[1] and s[2] > e[2]) then
+    local t = s
+    s = e
+    e = t
+  end
+  return { s = s, e = e }
+end
+
+---Get text from range.
+---@param range vim.snippet.Range
+---@return string
+local function get_text_by_range(range)
+  local lines = vim.api.nvim_buf_get_lines(0, range.s[1], range.e[1] + 1, false)
+  if range.s[1] ~= range.e[1] then
+    lines[1] = string.sub(lines[1], range.s[2] + 1)
+    lines[#lines] = string.sub(lines[#lines], 1, range.e[2])
+  else
+    lines[1] = string.sub(lines[1], range.s[2] + 1, range.e[2])
+  end
+  return table.concat(lines, '\n')
+end
+
+---Traverse snippet ast.
+---The `context.range` is calculated by `tostring(node)`, so it can be used only for processing before snippet insertion.
+---@param snippet_node table
+---@param callback fun(node: table, context: vim.snippet.TraverseContext): table|nil
+local function traverse(snippet_node, callback)
```

Comment:
👀  https://github.com/neovim/neovim/pull/25154 @llllvvuu

Reply:
oh yay another use case! just left a status update on that PR